% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/VarGen.R
\name{vargen_custom}
\alias{vargen_custom}
\title{Get variants related to specific genes.}
\usage{
vargen_custom(vargen_dir, gene_ids, fantom_corr = 0.25, outdir = "./",
  gtex_tissues, gwas_traits, gene_mart, snp_mart, verbose = FALSE)
}
\arguments{
\item{vargen_dir}{directory with the following file (can be generated with
\code{\link{vargen_install}})}

\item{gene_ids}{list of ensembl gene IDs of interest}

\item{fantom_corr}{the minimum correlation (z-score) to consider a FANTOM5
enhancer/gene association valid (default: 0.25).
A z-score greater than 0 represents an element greater than the mean, this
means that this association has more correlation than random motifs.}

\item{outdir}{the output directory, some files will be written during the
running of this function}

\item{gtex_tissues}{a vector containing the name of the "signif_variant_gene_pairs.txt.gz"
files. Output from \code{\link{select_gtex_tissues}} can be used.}

\item{gwas_traits}{a vector with the trait of interest (as characters). The list
of available traits can be obtained with \code{\link{list_gwas_traits}}}

\item{gene_mart}{optional, a connection to ensembl gene mart, can be created
using \code{\link{connect_to_gene_ensembl}} (If missing this function will be
used to create the connection).}

\item{snp_mart}{optional, a connection to ensembl snp mart, can be created
using \code{\link{connect_to_snp_ensembl}} (If missing this function will be
used to create the connection).}

\item{verbose}{if TRUE, will print progress messages (default: FALSE)}
}
\value{
a data.frame with the variants fetched from OMIM, FANTOM5, GTEx and GWAS.
The data.frame will contain the following columns:
\itemize{
  \item chr (chromosome)
  \item pos (position of the variant)
  \item rsid (variant ID)
  \item ensembl_gene_id ("gene id" of the gene associated with the variant)
  \item hgnc_symbol ("hgnc symbol" of the gene associated with the variant)
  \item source ("omim", "fantom5", "gtex" or "gwas")
}
}
\description{
Alternative to vargen_pipeline for a set of custom genes.
The variants are fetched from the following sources:
\itemize{
  \item get variants on the genes
  \item get the variants on the enhancers / promoters of the genes
  \item GTEx: get variants impacting the expression of the genes in specific tissues.
}
}
\examples{
# Simple query
vargen_install(install_dir = "./vargen_data/")

vargen_custom(vargen_dir = "./vargen_data/",
              gene_ids = c("ENSG00000166603", "ENSG00000155846"),
              outdir = "./", verbose = TRUE)

# With gwas and gtex
adipose_tissues <- select_gtex_tissues(gtex_dir = "./vargen_data/GTEx_Analysis_v8_eQTL/",
                                       tissues_query = "adipose")
vargen_custom(vargen_dir = "./vargen_data/",
              gene_ids = c("ENSG00000166603", "ENSG00000155846"),
              outdir = "./", verbose = TRUE,
              gtex_tissues = adipose_tissues,
              gwas_traits = "Obesity")
}
